Fix bug in detaching disks when node is powered-off but still in inventory in vsphere CP

w-leads opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
vSphere cloud provider fails to detach disks from powered-off node when the node is still in vCenter inventory. Due to this behavior, the disk needs to be detached manually before powering it on.
This issue is caused because node information is removed from nodemanager when kubernetes detects powered-off node though the information is required when VCP attempts to detach disks afterward. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

w-leads committed:
Fix bug in detaching disks in vsphere cp

VCP fails to detach disk from powered-off node which is still in
vcenter inventory. That's caused because VCP tries to detach disks
after nodemanager has unregistered the node information.

w-leads commented:
/sig vmware

dims commented:
/ok-to-test

w-leads commented:
/retest

w-leads commented:
/assign @abrarshivani

abrarshivani commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -780,12 +781,26 @@ func (vs *VSphere) DetachDisk(volPath string, nodeName k8stypes.NodeName) error
-				glog.Infof("Node %q does not exist, disk %s is already detached from node.", convertToString(nodeName), volPath)
+		if err == vclib.ErrNoVMFound {
+			// If node doesn't exist, discover the node to check if node is still in vcenter inventory
+			node, err := vs.nodeManager.GetRemovedNode(nodeName)
```
I think this will not work if controller-manager restarts since this information will vanish. Can you please confirm this?

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -45,10 +45,12 @@ type NodeManager struct {
 	nodeInfoMap map[string]*NodeInfo
 	// Maps node name to node structure
 	registeredNodes map[string]*v1.Node
+	removedNodes    map[string]*v1.Node
```
Node information will still stay even if all the volumes that are attached by the cluster are detached. I would suggest to remove this information when not needed.

w-leads committed:
Revert "Fix bug in detaching disks in vsphere cp"

This reverts commit bd618fbafe92fe54acbfa63076cbec81474069d6.

w-leads committed:
Fix bug in detaching disks in vsphere cp

VCP fails to detach disk from powered-off node which is still in
vcenter inventory. That's caused because VCP tries to detach disks
after nodemanager has unregistered the node information.

w-leads commented:
@abrarshivani Would you please take another look at code?
I reverted previous commit, and took the different approach by keeping attached disk information instead of removed node info.
To solve controller-manager restart case, the attached disks information are updated in VSphere.DisksAreAttached which is called almost every minute. 
Thank you,

abrarshivani commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1188,6 +1208,14 @@ func (vs *VSphere) NodeDeleted(obj interface{}) {
+	disks, err := vs.nodeManager.GetAttachedDisks(convertToK8sType(node.ObjectMeta.Name))
+	if err == nil {
+		for _, disk := range disks {
+			if err := vs.DetachDisk(disk, convertToK8sType(node.ObjectMeta.Name)); err != nil {
```
DetachDisk won't be retried if it fails (example: In case of network issue to vCenter) and still the volume will be attached to power off node. When kubernetes will retry to detach disk from the deleted node it will fail since node will be unregistered here.

abrarshivani commented:
@w-leads Thanks for fixing this. I think the right approach would be to report nodeName as nodeUUID. This way vSphere Cloud Provider will not have any need to maintain the mapping of nodeName to v1.Node or keep the state of attached and detached disk.

w-leads commented:
@abrarshivani Thanks a lot for reviewing again!

So what's your saying is following, is my understanding correct?
- Register mapping of nodeName to VM UUID(or maybe NodeInfo struct?), as done in 1st commit but instead of nodeName
- When k8s tries to detach disks from powered-off node and gets some errors, detach disks using that info instead
- Compare nodeVolumes and the info in VSphere.DisksAreAttached, and remove the info if there's a node which is not in nodeVolumes. Or maybe other better way to do this...?

In this case, I have no idea to get the info again in case that controller-manager is restarted after the node is powered-off. This is because only VSphere.NodeAdded, NodeDeleted receive v1.Node which contains UUID. Please correct me if I'm wrong

w-leads committed:
Retry to detach disks from deleted nodes in VCP

w-leads commented:
@abrarshivani I made a commit and now detaching disks from removed node will be retried when it fails. If this attached-disks-info method is okay for now, would you take another look please?

abrarshivani commented:
@w-leads Sounds good to me. Will take another look. Can you please create an issue to track Virtual Machine UUID as Nodename approach?

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -257,6 +260,38 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
+func (nm *NodeManager) GetAttachedDisks(nodeName k8stypes.NodeName) []string {
+	nm.attachedDisksLock.Lock()
+	attachedDisks := nm.attachedDisks[convertToString(nodeName)]
+	nm.attachedDisksLock.Unlock()
```
I think attachedDisks needs to be protected with locks. Since this is the same object which will be accessed by  RegisterAttachedDisk and UnRegisterAttachedDisk

abrarshivani commented on pkg/cloudprovider/providers/vsphere/vsphere_util.go:
```diff
@@ -380,30 +380,26 @@ func convertVolPathToDevicePath(ctx context.Context, dc *vclib.Datacenter, volPa
-		if err != nil {
-			return nil, err
-		}
+func (vs *VSphere) convertVolPathsToDevicePaths(ctx context.Context, nodeName k8stypes.NodeName, volPaths []string) ([]string, error) {
```
Can we go with the earlier way where we accept nodeVolumes here? ErrVMNotFound error can be handled here.

w-leads replied:
I suppose this change is necessary.
The function before this change received volumes on all nodes, and if any of nodes was not found, the function and VSphere.DisksAreAttached returned error. However attached-disks information needs to be re-registered in DisksAreAttached in case of controller manager restart. That's why I changed this func to handle volumes on a single node so that DisksAreAttached could just ignore removed nodes instead of returning error before reregistration.
Also, DisksAreAttached has a comment which says "// 5b. If VMs are removed from vSphere inventory they are ignored." So I suppose DisksAreAttached shouldn't give error at this point.

w-leads committed:
Use defer to unlock attached disks in vcp

abrarshivani commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1191,7 +1213,25 @@ func (vs *VSphere) NodeDeleted(obj interface{}) {
+	go func() {
+		nodeName := convertToK8sType(node.ObjectMeta.Name)
+		for i := 0; i < CheckDisksOnNodeDeletedMaxRetry; i++ {
+			disks := vs.nodeManager.GetAttachedDisks(nodeName)
```
I think if controller manager is restarted in somewhere at this point then detach disk will still fail because detach disk won't be able to find the *v1.Node for the respective node.

w-leads replied:
Thanks for reviewing. I think we could manage that case by keeping removed-node information as I did in 1st commit in addition to attached-disks info, until we rewrite the code using vm uuid as nodename. However this could happen when 3 unusual things happen at the same time: (1) node shutdown on which some containers running (2) problem connecting to vcenter (3) controller mangers restarted in other nodes.
I'm wondering if we could go with current way instead of adding some complexity to solve it for now.

SandeepPissay replied:
I agree with Abrar. Even though the 3 things you mentioned may not happen at the same time, this is not a reliable solution. Instead of pushing this change now, why not have a complete solution? 

Any reason why you want to push this change soon?

yastij commented:
Please see #58635 as this is not only related to Vsphere.

yastij commented:
cc @jingxu97

jingxu97 commented:
@w-leads Thanks for the PR. Could you please add a design description (how the issue is fixed) into the PR so it is easier to review? Thanks!

w-leads commented:
What I'd like to do in this PR is to make sure disks properly detached from powered-off nodes. This worked in v1.8.x, but it doesn't in v1.9.x or later because data necessary to call vSphere API is gone as soon as node is powered off. Also disks are expected to be unmounted from powered off nodes according to this document. So this should be fixed.
https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/high-availability.html

To make this happen, we have 2 different approaches.
(1) Keep additional info to detach disks from powered-off nodes
This is the way taken in current changes. Since nodemanager throws away node information when the node is deleted(sorry I'm still looking through related issues/pr), some other info is needed to detach disks. However, we need to take into consideration how to get the info back if controller manager restarts because the node is already gone.

(2) Report VM UUID as node name instead of hostname
Currently we're using hostname as a node name, and VCP needs a mapping to translate it to VM UUID in nodemanager so that it can find a virtual machine. However VCP could find VM directly by reporting UUID as node name, and VCP wouldn't need to store node info in itself and to handle controller crash case as the 1st approach. That being said, it'll have changes in a wide range of source code and would take long. That's why we didn't take this way here and now, and created another ticket #63987

I'd like to keep this PR as small as possible, and take next step to improve it by 2nd approach after that. So I cannot say that this way is good and error tolerant at this point but if it's acceptable I'd like to proceed with this by following changes. 

- Keep mapping info from nodename to its attached disks in nodemanager
  - Register it when a new disk attached in VSphere.AttachDisk, and unresiter it when it's detached in VSphere.DetachDisk
  - To recover the info from controller manager restart, register it if it's not registered yet in VSphere.DisksAreAttached
- When node is deleted, try to detach disks using the mapping info in VSphere.NodeDeleted. Then remove node info once it's done or abort detach operation if it's failed 10 times,
- Disks would remain attached if ...
  - vSphere is unreachable for some reason while retrying 10 times in VSphere.NodeDeleted
  - controller manager is restarted while processing VSphere.NodeDeleted

Please correct me if I'm wrong with something or taking inappropriate approach.
Thank you!

SandeepPissay commented:
@abrarshivani and I discussed a solution to this problem. Here's the solution:

vSphere cloud provider(VCP) would store the node to VM UUID mapping in API server during node membership events. This can be maintained in a config map or custom resource definition(CRD). On node unregister event, we do not remove the mapping. 

When VCP gets a callback for detaching disks (since the node went down and K8s scheduled the stateful pod on another node). VCP will look into the cache to find the node details. If not found, it will lookup into the config map or CRD to get the VM UUID. Once we have the VM UUID, we can find the VSphere instance and detach the disks from the node VM.

In order for this to work, VCP would now need a service account.

w-leads committed:
Revert "Use defer to unlock attached disks in vcp"

This reverts commit 680ab089ce491f27c3602202761d2dd2625bec90.

w-leads committed:
Revert "Retry to detach disks from deleted nodes in VCP"

This reverts commit 0738258143024a28917c4626f6474ffbee552a55.

w-leads committed:
Revert "Fix bug in detaching disks in vsphere cp"

This reverts commit eea9efdd61a25c3e74dbc54e11aae711ca499c85.

w-leads commented:
@SandeepPissay @abrarshivani Thank you for discussing and sharing it. It sounds nice, I'll try to fix the issue using that way. We don't remove the mapping when node is unregistered, but still we need to remove it when the mapping is no longer necessary, is this right?

w-leads committed:
Fix bug in detaching disks in vsphere cp

VCP fails to detach disk from powered-off node which is still in
vcenter inventory. That's caused because VCP tries to detach disks
after nodemanager has unregistered the node information.

w-leads commented:
@abrarshivani @SandeepPissay If a node is powered off and then controller manager is crashed, restarted controller manager doesn't know disks which were attached to the powered-off node, and VSphere.DetachDisk is never called. I suppose that, to handle this case properly, VCP needs to store info about which disks are attached to which nodes in another config map, and call DetachDisk func when node is deleted. I feel this is not a right way but cannot come up with a better way. Would you please let me know that if you have an idea?

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -260,6 +258,97 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
+	if ok {
+		return vmUUID, nil
+	}
+	configMap, err := nm.getVMUUIDConfigMap()
```
Add a log that VM uuid was not cache

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -260,6 +258,97 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
+		glog.Errorf("vm uuid for %s not found in config map", nodeName)
+		return "", err
+	}
+	return vmUUID, nil
```
Add a log for nodeName and uuid which is found before returning

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -260,6 +258,97 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
+	}
+	_, err := nm.kubeClient.CoreV1().ConfigMaps(nm.vmUUIDConfigMapNamespace).Update(configMap)
+	if err != nil {
+		return err
```
Log err here

abrarshivani commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -226,6 +229,19 @@ func init() {
+	// Only on controller node it is required to register listeners.
+	// Register callbacks for node updates
+	// EventHandler needs to be added once kubeClient is provided to nodemanager
+	if vs.informerFactory != nil {
```
Why not keep the earlier logic where node events methods are registered in SetInformers?

w-leads replied:
Thanks for your review! What happens if we don't do this in Initialize method is that NodeAdded was called before kubeClient is passed to NodeManager, and NodeManager couldn't access configmaps.
We can keep it in SetInformers by adding some retry code in case that kubeClient is not yet to be given, but I feel it's more simple just to make sure NodeAdded/Deleted always called after kubeClient is provided instead. (Also I added L236 comment)

abrarshivani commented:
@w-leads Thanks for updating the solution. Deleting the node uuid mapping will not be handled in this PR, right?
Can you post the workflows that you have tested with this change? 

I will be on PTO starting tomorrow, so @divyenpatel or @BaluDontu can take a look at this, assigning @divyenpatel for now.
/assign @divyenpatel

w-leads committed:
Add logs for vm uuid configmap in vsphere cp

w-leads committed:
Assume disk is detached if node is not in vsphere inventory

w-leads commented:
Hi @divyenpatel @BaluDontu Would you please take a look at this PR?
Thank you

divyenpatel commented:
sure @w-leads we will take a look.

yastij commented:
cc @jingxu97 - might this create a race condition with the node shutdown taint (which planned for 1.12)

w-leads commented:
@yastij @jingxu97 This change won't be necessary when node shutdown taint is implemented in v1.12. If my understand is correct, when VM is in shutdown(powered-off) state, and InstanceShutdownByProviderID(we need to implement this in VCP) returns true, node controller will add shutdown taint, and node object will never be deleted from k8s cluster. Since node won't be deleted, NodeDeleted func in VCP won't be called, and node information which is necessary to detach disks from powered-off node won't be deleted too. So the issue to be fixed in this PR won't happen.
So I'm wondering if we could add changes in this PR only to prior versions v1.9 to v1.11. Or since this change won't make any conflict with shutdown taint and will be just unnecessary, we can keep this until some code to handle shutdown taint is implemented, and then remove this change.

SandeepPissay commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -233,6 +236,19 @@ func init() {
 
 // Initialize passes a Kubernetes clientBuilder interface to the cloud provider
 func (vs *VSphere) Initialize(clientBuilder controller.ControllerClientBuilder) {
```
Is this method getting invoked in Kubelet? If no, then we should fail the initialization if vs.informerFactory is not set.

Can you confirm if Initialize is invoked after SetInformers?

You are also introducing a vsphere-cloud-provider service account, wouldn't this break Kubernetes upgrade? How do you plan to address upgrade?

gnufied replied:
I think we will have to install corresponding rbac policies as defaults.

gnufied replied:
This does not get called from kubelet. Also `Initialize` is called after `SetInformers`, so I think we should be safe. Having said that, I think the code should work if it can't find configmap as well and revert to old behaviour when configmap isn't found.

abrarshivani commented on pkg/cloudprovider/providers/vsphere/nodemanager.go:
```diff
@@ -260,6 +258,101 @@ func (nm *NodeManager) removeNode(node *v1.Node) {
+		return nil, err
+	}
+	if configMap.Data == nil {
+		nm.vmUUIDConfigMapCache = make(map[string]string)
```
Giving configmap read and write permissions to vSphere Cloud Provider after launching k8s cluster will fail the volume operations. This happens due to node events are occurred before giving permissions and nodeName and vmUUID mapping is not cached at that time. Can you fix this?

gnufied replied:
Installing the RBAC permissions via bootstrap policies should fix that right?

gnufied commented:
@w-leads does this PR really needs to be against master because IIRC master isn't affected by this bug.
